Improve numeric stability for LayerNorm on CPU

xiaomengy opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40308 Improve numeric stability for LayerNorm on CUDA
* **#40307 Improve numeric stability for LayerNorm on CPU**

Improve numeric stability for LayerNorm on CPU.

This diff will slightly decrease the performance of LayerNorm but increase the numerical stability especially when the variance is small.

Example.

M = 1024, N = 32768
input =  torch.randn(M, N, dtype=torch.float) * 0.01 + 100.0

Previously in this case, the computed variance will be all 0s.
After this diff, the variance will be

[1.002566e-04, 9.923905e-05, 1.014871e-04, ..., 1.015831e-04, 1.000394e-04, 1.009022e-04]

This is is close to 1e-4 as we expected. The max relative difference compare to double result is 0.0005683.

The average running time for F.layer_norm with input above on one devgpu before this PR is about 99.68ms, and after this PR is about 101.32ms. So the influence on performance should be limited.

Differential Revision: [D21993646](https://our.internmc.facebook.com/intern/diff/D21993646/)

xiaomengy committed:
Improve numeric stability for LayerNorm on CPU

Improve numeric stability for LayerNorm on CPU.

This diff will slightly decrease the performance of LayerNorm but increase the numerical stability especially when the variance is small.

Example.

M = 1024, N = 32768
input =  torch.randn(M, N, dtype=torch.float) * 0.01 + 100.0

Previously in this case, the computed variance will be all 0s.
After this diff, the variance will be

[1.002566e-04, 9.923905e-05, 1.014871e-04, ..., 1.015831e-04, 1.000394e-04, 1.009022e-04]

This is is close to 1e-4 as we expected. The max relative difference compare to double result is 0.0005683.

Differential Revision: [D21993646](https://our.internmc.facebook.com/intern/diff/D21993646/)

[ghstack-poisoned]

BIT-silence commented:
link https://github.com/pytorch/pytorch/issues/40302

xiaomengy committed:
Update on "Improve numeric stability for LayerNorm on CPU"


Improve numeric stability for LayerNorm on CPU.

This diff will slightly decrease the performance of LayerNorm but increase the numerical stability especially when the variance is small.

Example.

M = 1024, N = 32768
input =  torch.randn(M, N, dtype=torch.float) * 0.01 + 100.0

Previously in this case, the computed variance will be all 0s.
After this diff, the variance will be

[1.002566e-04, 9.923905e-05, 1.014871e-04, ..., 1.015831e-04, 1.000394e-04, 1.009022e-04]

This is is close to 1e-4 as we expected. The max relative difference compare to double result is 0.0005683.

The average running time for F.layer_norm with input above on one devgpu before this PR is about 124ms, and after this PR is about 140ms. So the influence on performance should be limited.

Differential Revision: [D21993646](https://our.internmc.facebook.com/intern/diff/D21993646/)

[ghstack-poisoned]

xiaomengy committed:
Update on "Improve numeric stability for LayerNorm on CPU"


Improve numeric stability for LayerNorm on CPU.

This diff will slightly decrease the performance of LayerNorm but increase the numerical stability especially when the variance is small.

Example.

M = 1024, N = 32768
input =  torch.randn(M, N, dtype=torch.float) * 0.01 + 100.0

Previously in this case, the computed variance will be all 0s.
After this diff, the variance will be

[1.002566e-04, 9.923905e-05, 1.014871e-04, ..., 1.015831e-04, 1.000394e-04, 1.009022e-04]

This is is close to 1e-4 as we expected. The max relative difference compare to double result is 0.0005683.

The average running time for F.layer_norm with input above on one devgpu before this PR is about 124ms, and after this PR is about 140ms. So the influence on performance should be limited.

Differential Revision: [D21993646](https://our.internmc.facebook.com/intern/diff/D21993646/)

[ghstack-poisoned]

BIT-silence commented on aten/src/ATen/native/cuda/layer_norm_kernel.cu:
```diff
@@ -328,8 +335,26 @@ void LayerNormBackwardKernelImplInternal(
+
+  if (M == 0) {
+    if (dgamma->defined()) {
+      T* dgamma_data = dgamma->data_ptr<T>();
```
Changed back to at::zeros outside.

BIT-silence commented on aten/src/ATen/native/layer_norm.cpp:
```diff
@@ -27,13 +27,12 @@ std::tuple<Tensor, Tensor, Tensor> layer_norm_cpu(
-  if (M > 0) {
-    LayerNormKernel(kCPU, X, gamma, beta, M, N, eps, &Y, &mean, &rstd);
-  }
+  LayerNormKernel(
```
Reverted.

xiaomengy committed:
Update on "Improve numeric stability for LayerNorm on CPU"


Improve numeric stability for LayerNorm on CPU.

This diff will slightly decrease the performance of LayerNorm but increase the numerical stability especially when the variance is small.

Example.

M = 1024, N = 32768
input =  torch.randn(M, N, dtype=torch.float) * 0.01 + 100.0

Previously in this case, the computed variance will be all 0s.
After this diff, the variance will be

[1.002566e-04, 9.923905e-05, 1.014871e-04, ..., 1.015831e-04, 1.000394e-04, 1.009022e-04]

This is is close to 1e-4 as we expected. The max relative difference compare to double result is 0.0005683.

The average running time for F.layer_norm with input above on one devgpu before this PR is about 124ms, and after this PR is about 140ms. So the influence on performance should be limited.

Differential Revision: [D21993646](https://our.internmc.facebook.com/intern/diff/D21993646/)

[ghstack-poisoned]

BIT-silence commented on aten/src/ATen/native/cuda/layer_norm_kernel.cu:
```diff
@@ -268,6 +268,9 @@ void LayerNormKernelImplInternal(
   DCHECK_EQ(X.numel(), M * N);
   DCHECK(!gamma.defined() || gamma.numel() == N);
   DCHECK(!beta.defined() || beta.numel() == N);
```
Actually they were changed in the next PR.

xiaomengy committed:
Update on "Improve numeric stability for LayerNorm on CPU"


Improve numeric stability for LayerNorm on CPU.

This diff will slightly decrease the performance of LayerNorm but increase the numerical stability especially when the variance is small.

Example.

M = 1024, N = 32768
input =  torch.randn(M, N, dtype=torch.float) * 0.01 + 100.0

Previously in this case, the computed variance will be all 0s.
After this diff, the variance will be

[1.002566e-04, 9.923905e-05, 1.014871e-04, ..., 1.015831e-04, 1.000394e-04, 1.009022e-04]

This is is close to 1e-4 as we expected. The max relative difference compare to double result is 0.0005683.

The average running time for F.layer_norm with input above on one devgpu before this PR is about 124ms, and after this PR is about 140ms. So the influence on performance should be limited.

Differential Revision: [D21993646](https://our.internmc.facebook.com/intern/diff/D21993646/)

[ghstack-poisoned]

xiaomengy committed:
Update on "Improve numeric stability for LayerNorm on CPU"


Improve numeric stability for LayerNorm on CPU.

This diff will slightly decrease the performance of LayerNorm but increase the numerical stability especially when the variance is small.

Example.

M = 1024, N = 32768
input =  torch.randn(M, N, dtype=torch.float) * 0.01 + 100.0

Previously in this case, the computed variance will be all 0s.
After this diff, the variance will be

[1.002566e-04, 9.923905e-05, 1.014871e-04, ..., 1.015831e-04, 1.000394e-04, 1.009022e-04]

This is is close to 1e-4 as we expected. The max relative difference compare to double result is 0.0005683.

The average running time for F.layer_norm with input above on one devgpu before this PR is about 124ms, and after this PR is about 140ms. So the influence on performance should be limited.

Differential Revision: [D21993646](https://our.internmc.facebook.com/intern/diff/D21993646/)

[ghstack-poisoned]

xiaomengy committed:
Update on "Improve numeric stability for LayerNorm on CPU"


Improve numeric stability for LayerNorm on CPU.

This diff will slightly decrease the performance of LayerNorm but increase the numerical stability especially when the variance is small.

Example.

M = 1024, N = 32768
input =  torch.randn(M, N, dtype=torch.float) * 0.01 + 100.0

Previously in this case, the computed variance will be all 0s.
After this diff, the variance will be

[1.002566e-04, 9.923905e-05, 1.014871e-04, ..., 1.015831e-04, 1.000394e-04, 1.009022e-04]

This is is close to 1e-4 as we expected. The max relative difference compare to double result is 0.0005683.

The average running time for F.layer_norm with input above on one devgpu before this PR is about 99.68ms, and after this PR is about 101.32ms. So the influence on performance should be limited.

Differential Revision: [D21993646](https://our.internmc.facebook.com/intern/diff/D21993646/)

[ghstack-poisoned]

xiaomengy commented:
https://github.com/pytorch/pytorch/pull/59987 should fix the issue
